boa_engine/src/bytecompiler/jump_control.rs

```diff
@@ -0,0 +1,366 @@
+        self.jump_info.push(new_info);
+    }
+
+    pub(crate) fn pop_switch_control_info(&mut self) {
+        let info = self.jump_info.pop().expect("no jump information found");
+
+        assert!(info.kind == JumpControlInfoKind::Switch);
+
+        for label in info.breaks {
+            self.patch_jump(label);
+        }
+    }
+
+    pub(crate) fn push_try_control_info(&mut self, has_finally: bool) {
+        if !self.jump_info.is_empty() {
+            let start_address = self
+                .jump_info
+                .last()
+                .expect("no jump information found")
+                .start_address();
+
+            let new_info = JumpControlInfo::default()
+                .with_kind(JumpControlInfoKind::Try)
+                .with_start_address(start_address)
+                .with_has_finally(has_finally);
+
+            self.jump_info.push(new_info);
+        }
+    }
+
+    pub(crate) fn pop_try_control_info(&mut self, finally_start_address: Option<u32>) {
+        if !self.jump_info.is_empty() {
+            let mut info = self.jump_info.pop().expect("no jump information found");
+
+            assert!(info.kind == JumpControlInfoKind::Try);
+
+            let mut breaks = Vec::with_capacity(info.breaks.len());
+
+            if let Some(finally_start_address) = finally_start_address {
+                for label in info.try_continues {
+                    if label.index < finally_start_address {
+                        self.patch_jump_with_target(label, finally_start_address);
+                    } else {
+                        self.patch_jump_with_target(label, info.start_address);
+                    }
+                }
+
+                for label in info.breaks {
+                    if label.index < finally_start_address {
+                        self.patch_jump_with_target(label, finally_start_address);
+                        let Label { mut index } = label;
+                        index -= size_of::<Opcode>() as u32;
+                        index -= size_of::<u32>() as u32;
+                        breaks.push(Label { index });
+                    } else {
+                        breaks.push(label);
+                    }
+                }
+                if let Some(jump_info) = self.jump_info.last_mut() {
+                    jump_info.breaks.append(&mut breaks);
+                }
+            } else if let Some(jump_info) = self.jump_info.last_mut() {
+                jump_info.breaks.append(&mut info.breaks);
+                jump_info.try_continues.append(&mut info.try_continues);
+            }
+        }
+    }
+
+    pub(crate) fn push_labelled_block_control_info(&mut self, label: Sym, start_address: u32) {
+        let new_info = JumpControlInfo::default()
+            .with_kind(JumpControlInfoKind::LabelledBlock)
+            .with_label(Some(label))
+            .with_start_address(start_address);
+        self.jump_info.push(new_info);
+    }
```

Comment:
I'm seeing that some fields on the `JumpControlInfo` are only used on some `JumpControlInfoKind`s; `jump_info.breaks` only makes sense on loops and blocks, and `has_finally` only makes sense on try blocks. Could those fields be moved into the `JumpControlInfoKind` itself?

Reply:
I think it could if it was switched over to one bitflag. I was playing around with the idea when I was working on the changes initially but had decided to not do it at the time. I tried it out on the most recent commit, let me know what you think. 😄

Reply:
Looks good! I was thinking more about using an ADT to represent the types of control infos, but on retrospect I think that'll require a lot more design work and I don't know if it'll be worth it.